[jit] clean up import_source

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24282 [jit] clean up import_source**
* #24281 [jit] class_table_ to deps_table_
* #24280 [jit] make FunctionType a NamedType
* #24279 [jit] make NamedType an interface
* #24278 [jit] simplify NamedType interface

This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

Differential Revision: [D16800562](https://our.internmc.facebook.com/intern/diff/D16800562)

suo committed:
[jit] clean up import_source

This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

suo committed:
Update on "[jit] clean up import_source"


This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

ZolotukhinM reviewed (approved):
Looks good (I've already reviewed this previously)! One thing I suggest to add now is a reference to the original commit that got reverted and a brief description of what happened and how it's mitigated now.

suo commented:
Sure, can do!

suo committed:
Update on "[jit] clean up import_source"


This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

suo committed:
Update on "[jit] clean up import_source"


This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

suo committed:
Update on "[jit] clean up import_source"


This moves a test from Python to cpp, and in doing so lets us clean up a
bunch of otherwise unused code.

Differential Revision: [D16800562](https://our.internmc.facebook.com/intern/diff/D16800562)

suo commented:
crap, I totally forgot to to this @ZolotukhinM. Hopefully this comment thread will be enough for posterity :( 

The original stack that this is re-landing is at https://github.com/pytorch/pytorch/pull/23846. It was reverted due to a windows build breakage; I am relanding after fixing.
